stdlib/public/Concurrency/Actor.cpp

```diff
@@ -935,356 +1034,370 @@ static void traceJobQueue(DefaultActorImpl *actor, Job *first) {
   });
 }
 
-void DefaultActorImpl::giveUpThread(RunningJobInfo runner) {
-  SWIFT_TASK_DEBUG_LOG("giving up thread for actor %p", this);
-  auto oldState = CurrentState.load(std::memory_order_acquire);
-  assert(oldState.Flags.isAnyRunningStatus());
+static inline void traceActorStateTransition(DefaultActorImpl *actor,
+    ActiveActorStatus oldState, ActiveActorStatus newState) {
 
-  auto firstNewJob = preprocessQueue(oldState.FirstJob, JobRef(), nullptr);
-  traceJobQueue(this, firstNewJob);
+  SWIFT_TASK_DEBUG_LOG("Actor %p transitioned from %zx to %zx (%s)\n", actor,
+    oldState.getOpaqueFlags(), newState.getOpaqueFlags(), __FUNCTION__);
+  concurrency::trace::actor_state_changed(actor,
+        newState.getFirstJob().getRawJob(), newState.getFirstJob().needsPreprocessing(),
+        newState.getOpaqueFlags());
+}
 
-  _swift_tsan_release(this);
-  while (true) {
-    // In Zombie_ReadyForDeallocation state, nothing else should
-    // be touching the atomic, and there's no point updating it.
-    if (oldState.Flags.getStatus() == Status::Zombie_ReadyForDeallocation) {
-      deallocateUnconditional();
+void DefaultActorImpl::destroy() {
+  auto oldState = CurrentState.load(std::memory_order_relaxed);
+  // Tasks on an actor are supposed to keep the actor alive until they start
+  // running and we can only get here if ref count of the object = 0 which means
+  // there should be no more tasks enqueued on the actor.
+  assert(!oldState.getFirstJob() && "actor has queued jobs at destruction");
+
+  if (oldState.isIdle()) {
       return;
-    }
+  }
+  assert(oldState.isRunning() && "actor scheduled but not running at destruction");
+}
+
+void DefaultActorImpl::deallocate() {
+  // If we're running, mark ourselves as ready for deallocation but don't
+  // deallocate yet. When we stop running the actor - at unlock() time - we'll
+  // do the actual deallocation.
+  //
+  // If we're idle, just deallocate immediately
+  auto oldState = CurrentState.load(std::memory_order_relaxed);
+  while (oldState.isRunning()) {
+    auto newState = oldState.withZombie_ReadyForDeallocation();
+    if (CurrentState.compare_exchange_weak(oldState, newState,
+                                           std::memory_order_relaxed,
+                                           std::memory_order_relaxed))
+      return;
+  }
 
-    // In Zombie_Latching state, we should try to update to Idle;
-    // if we beat the releasing thread, it'll deallocate.
-    // In Running state, we may need to schedule a processing job.
+  assert(oldState.isIdle());
```

Comment:
I wonder if some of these asserts should be upgraded to fire in release builds too. This one, for example, seems like something cheap and important enough to be unconditional. Doesn't have to go into this PR, just something to ponder.